daemon/cluster/listen_addr_linux.go

```diff
@@ -0,0 +1,75 @@
+// +build linux
+
+package cluster
+
+import (
+	"net"
+
+	"github.com/vishvananda/netlink"
+)
+
+func (c *Cluster) resolveSystemAddr() (net.IP, error) {
+	// Use the system's only device IP address, or fail if there are
+	// multiple addresses to choose from.
+	interfaces, err := netlink.LinkList()
+	if err != nil {
+		return nil, err
+	}
+
+	var systemAddr net.IP
+
+	for _, intf := range interfaces {
+		// Skip non device or inactive interfaces
+		if intf.Type() != "device" || intf.Attrs().Flags&net.FlagUp == 0 {
+			continue
+		}
+
+		addrs, err := netlink.AddrList(intf, netlink.FAMILY_ALL)
+		if err != nil {
+			continue
+		}
+
+		var interfaceAddr4, interfaceAddr6 net.IP
+
+		for _, addr := range addrs {
+			ipAddr := addr.IPNet.IP
+
+			// Skip loopback and link-local addresses
+			if !ipAddr.IsGlobalUnicast() {
+				continue
+			}
+
+			if ipAddr.To4() != nil {
+				if interfaceAddr4 != nil {
+					return nil, errMultipleIPs
+				}
+				interfaceAddr4 = ipAddr
+			} else {
+				if interfaceAddr6 != nil {
+					return nil, errMultipleIPs
+				}
+				interfaceAddr6 = ipAddr
+			}
+		}
+
+		// In the case that this interface has exactly one IPv4 address
+		// and exactly one IPv6 address, favor IPv4 over IPv6.
+		if interfaceAddr4 != nil {
+			if systemAddr != nil {
+				return nil, errMultipleIPs
+			}
+			systemAddr = interfaceAddr4
+		} else if interfaceAddr6 != nil {
+			if systemAddr != nil {
+				return nil, errMultipleIPs
+			}
+			systemAddr = interfaceAddr6
+		}
+	}
+
+	if systemAddr == nil {
+		return nil, errNoIP
```

Comment:
We could do it, but wouldn't that be even more confusing to explain if a user actually ends is such a situation ?

In my view the autodetection is for covering the regular 99% of cases, where a physical device provides the connectivity to outside the host, and that was the reason behind this change.
Therefore it seems reasonable to me to have the autodetection fail if it cannot find an "eth" like device. Also I feel it is something people would find natural, and would be comfortable with it, in case of failure.

If we add self-healing logic, I am afraid it gets complicated, both to explain and even more to maintain. For example, I am not convinced we should choose a bridge interface as second external connectivity choice, rather than an ipvaln interface, for example. I feel it could easily take us into a rat hole of a priority list for backup interface types.

In summary, I'd prefer a simple predictable logic which covers the majority of the users' regular cases.

But of course if there is a strong argument in favor of a more intelligent autodetection mechanism and it turns out there are many use cases to support it, we should do it.